// nuxi/dist/chunks/generate.mjs
import { d as defineCommand, e as legacyRootDirArgs, f as envNameArgs, b as dotEnvArgs, l as logLevelArgs, c as cwdArgs } from '../shared/nuxi.DCVnYqfZ.mjs';
import buildCommand from './build.mjs';
import 'node:crypto';
import 'node:path';
import 'node:process';
import 'node:util';
import 'node:tty';
import 'node:url';
import '../shared/nuxi.Q_h4raRv.mjs';
import '../shared/nuxi.wIrae1ai.mjs';
import 'node:fs';
import 'node:assert';
import 'node:v8';
import 'node:fs/promises';
import '../shared/nuxi.Do0aYBCO.mjs';
import '../shared/nuxi.BxIDvNqq.mjs';
import 'node:module';
import '../shared/nuxi.C4l48OVX.mjs';
import '../shared/nuxi.D5KtpA0t.mjs';
import 'node:os';
import 'node:perf_hooks';
import 'node:vm';
import 'assert';
import 'fs';
import 'module';
import 'os';
import 'path';
import 'process';
import 'tty';
import 'url';
import 'util';
import 'v8';
import '../shared/nuxi.BetH2Ldg.mjs';
import '../shared/nuxi.8VrikUNY.mjs';

const generate = defineCommand({
  meta: {
    name: "generate",
    description: "Build Nuxt and prerender all routes"
  },
  args: {
    ...cwdArgs,
    ...logLevelArgs,
    preset: {
      type: "string",
      description: "Nitro server preset"
    },
    ...dotEnvArgs,
    ...envNameArgs,
    ...legacyRootDirArgs
  },
  async run(ctx) {
    ctx.args.prerender = true;
    await buildCommand.run(
      // @ts-expect-error types do not match
      ctx
    );
  }
});

export { generate as default };
