// nuxi/dist/chunks/typecheck.mjs
import process from 'node:process';
import { fileURLToPath } from 'node:url';
import { d as defineCommand, e as legacyRootDirArgs, b as dotEnvArgs, l as logLevelArgs, c as cwdArgs, O } from '../shared/nuxi.DCVnYqfZ.mjs';
import { l as loadKit, c as createJiti } from '../shared/nuxi.BxIDvNqq.mjs';
import { x as ve } from './main.mjs';
import { r as resolve } from '../shared/nuxi.Do0aYBCO.mjs';
import 'node:crypto';
import 'node:path';
import 'node:util';
import 'node:tty';
import 'node:module';
import '../shared/nuxi.C4l48OVX.mjs';
import '../shared/nuxi.D5KtpA0t.mjs';
import 'node:os';
import 'node:fs';
import 'node:assert';
import 'node:v8';
import 'node:perf_hooks';
import 'node:vm';
import 'assert';
import 'fs';
import 'module';
import 'os';
import 'path';
import 'process';
import 'tty';
import 'url';
import 'util';
import 'v8';
import 'child_process';
import 'stream';
import 'readline';

const typecheck = defineCommand({
  meta: {
    name: "typecheck",
    description: "Runs `vue-tsc` to check types throughout your app."
  },
  args: {
    ...cwdArgs,
    ...logLevelArgs,
    ...dotEnvArgs,
    ...legacyRootDirArgs
  },
  async run(ctx) {
    process.env.NODE_ENV = process.env.NODE_ENV || "production";
    const cwd = resolve(ctx.args.cwd || ctx.args.rootDir);
    const { loadNuxt, buildNuxt, writeTypes } = await loadKit(cwd);
    const nuxt = await loadNuxt({
      cwd,
      dotenv: { cwd, fileName: ctx.args.dotenv },
      overrides: {
        _prepare: true,
        logLevel: ctx.args.logLevel
      }
    });
    await writeTypes(nuxt);
    await buildNuxt(nuxt);
    await nuxt.close();
    const jiti = createJiti(cwd);
    const [resolvedTypeScript, resolvedVueTsc] = await Promise.all([
      jiti.esmResolve("typescript", { try: true }),
      jiti.esmResolve("vue-tsc/bin/vue-tsc.js", { try: true })
    ]);
    if (resolvedTypeScript && resolvedVueTsc) {
      await ve(fileURLToPath(resolvedVueTsc), ["--noEmit"], {
        throwOnError: true,
        nodeOptions: {
          stdio: "inherit",
          cwd
        }
      });
    } else {
      if (O) {
        await ve(
          "bun",
          "install typescript vue-tsc --global --silent".split(" "),
          {
            throwOnError: true,
            nodeOptions: { stdio: "inherit", cwd }
          }
        );
        await ve("bunx", "vue-tsc --noEmit".split(" "), {
          throwOnError: true,
          nodeOptions: {
            stdio: "inherit",
            cwd
          }
        });
      } else {
        await ve(
          "npx",
          "-p vue-tsc -p typescript vue-tsc --noEmit".split(" "),
          {
            throwOnError: true,
            nodeOptions: { stdio: "inherit", cwd }
          }
        );
      }
    }
  }
});

export { typecheck as default };
