// nuxi/dist/chunks/dev-child.mjs
import process from 'node:process';
import { d as defineCommand, e as legacyRootDirArgs, b as dotEnvArgs, f as envNameArgs, l as logLevelArgs, c as cwdArgs, R, a as logger } from '../shared/nuxi.DCVnYqfZ.mjs';
import { d as defu } from '../shared/nuxi.D2_bzAv0.mjs';
import { _ as _getDevServerOverrides, a as _getDevServerDefaults, c as createNuxtDevServer } from './dev2.mjs';
import { o as overrideEnv } from '../shared/nuxi.BetH2Ldg.mjs';
import { r as resolve } from '../shared/nuxi.Do0aYBCO.mjs';
import 'node:crypto';
import 'node:path';
import 'node:util';
import 'node:tty';
import 'node:url';
import 'node:events';
import 'fs';
import 'fs/promises';
import 'events';
import 'path';
import 'stream';
import 'os';
import '../shared/nuxi.DqkIr2yV.mjs';
import '../shared/nuxi.D4vggcPe.mjs';
import '../shared/nuxi.BxIDvNqq.mjs';
import 'node:module';
import '../shared/nuxi.C4l48OVX.mjs';
import '../shared/nuxi.D5KtpA0t.mjs';
import 'node:os';
import 'node:fs';
import 'node:assert';
import 'node:v8';
import 'node:perf_hooks';
import 'node:vm';
import 'assert';
import 'module';
import 'process';
import 'tty';
import 'url';
import 'util';
import 'v8';
import './index2.mjs';
import 'node:http';
import 'node:https';
import 'node:net';
import 'http';
import 'https';
import 'node:child_process';
import 'crypto';
import 'node:fs/promises';
import '../shared/nuxi.8VrikUNY.mjs';
import '../shared/nuxi.2PplrG5f.mjs';

const devChild = defineCommand({
  meta: {
    name: "_dev",
    description: "Run Nuxt development server (internal command to start child process)"
  },
  args: {
    ...cwdArgs,
    ...logLevelArgs,
    ...envNameArgs,
    ...dotEnvArgs,
    ...legacyRootDirArgs
  },
  async run(ctx) {
    if (!process.send && !R) {
      logger.warn("`nuxi _dev` is an internal command and should not be used directly. Please use `nuxi dev` instead.");
    }
    overrideEnv("development");
    const cwd = resolve(ctx.args.cwd || ctx.args.rootDir);
    const devContext = JSON.parse(process.env.__NUXT_DEV__ || "null") || {};
    function sendIPCMessage(message) {
      if (process.send) {
        process.send(message);
      } else {
        logger.info(
          "Dev server event:",
          Object.entries(message).map((e) => `${e[0]}=${JSON.stringify(e[1])}`).join(" ")
        );
      }
    }
    process.once("unhandledRejection", (reason) => {
      sendIPCMessage({ type: "nuxt:internal:dev:rejection", message: reason instanceof Error ? reason.toString() : "Unhandled Rejection" });
      process.exit();
    });
    const devServerOverrides = _getDevServerOverrides({
      public: devContext.public
    });
    const devServerDefaults = _getDevServerDefaults({
      hostname: devContext.hostname,
      https: devContext.proxy?.https
    }, devContext.publicURLs);
    const nuxtDev = await createNuxtDevServer({
      cwd,
      overrides: defu(ctx.data?.overrides, devServerOverrides),
      defaults: devServerDefaults,
      logLevel: ctx.args.logLevel,
      clear: !!ctx.args.clear,
      dotenv: { cwd, fileName: ctx.args.dotenv },
      envName: ctx.args.envName,
      port: process.env._PORT ?? void 0,
      devContext
    });
    nuxtDev.on("loading:error", (_error) => {
      sendIPCMessage({ type: "nuxt:internal:dev:loading:error", error: {
        message: _error.message,
        stack: _error.stack,
        name: _error.name,
        code: _error.code
      } });
    });
    nuxtDev.on("loading", (message) => {
      sendIPCMessage({ type: "nuxt:internal:dev:loading", message });
    });
    nuxtDev.on("restart", () => {
      sendIPCMessage({ type: "nuxt:internal:dev:restart" });
    });
    nuxtDev.on("ready", (payload) => {
      sendIPCMessage({ type: "nuxt:internal:dev:ready", port: payload.port });
    });
    await nuxtDev.init();
  }
});

export { devChild as default };
